components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/util/ResponseTypeHandlerUtil.java

```diff
@@ -253,9 +254,15 @@ public static AuthzCodeDO generateAuthorizationCode(OAuthAuthzReqMessageContext
         // set code issued time.this is needed by downstream handlers.
         oauthAuthzMsgCtx.setCodeIssuedTime(timestamp.getTime());
 
-        if (authorizationReqDTO.getUser() != null && authorizationReqDTO.getUser().isFederatedUser()) {
-            //if a federated user, treat the tenant domain as similar to application domain.
-            authorizationReqDTO.getUser().setTenantDomain(authorizationReqDTO.getTenantDomain());
+        AuthenticatedUser authenticatedUser = authorizationReqDTO.getUser();
+        if (authenticatedUser != null && authenticatedUser.isFederatedUser()) {
+            String appName = OAuth2Util.getServiceProvider(authorizationReqDTO.getConsumerKey()).getApplicationName();
+            // If federated role-based authorization is engaged skip overwriting the user tenant domain.
+            if (!IdentityUtil.getPropertyAsList(FIDP_ROLE_BASED_AUTHZ_APP_CONFIG).contains(appName)
+                    || authenticatedUser.getTenantDomain() == null) {
+                // If a federated user, treat the tenant domain as similar to the application domain.
+                authenticatedUser.setTenantDomain(authorizationReqDTO.getTenantDomain());
```

Comment:
Shouldn't this logic be in the framework?
ie. Shouldn't these be set as part of authentication logic?